sdks/java/core/src/main/java/com/google/cloud/dataflow/sdk/options/ProxyInvocationHandler.java

```diff
@@ -209,6 +261,69 @@ public int hashCode() {
   }
 
   /**
+   * Populate display data. See {@link HasDisplayData#populateDisplayData}.
+   */
+  private void populateDisplayData(DisplayData.Builder builder) {
+    Preconditions.checkState(canPopulateDisplayData(),
+        "Populating display data from deserialized PipelineOptions is not supported.");
+    Set<PipelineOptionsReflector.Property> props =
+        PipelineOptionsReflector.collectVisibleProperties(knownInterfaces);
+
+    Multimap<String, Class<?>> keyToInterfaceMap = buildOptionToInterfaceMap(props);
+
+    for (Map.Entry<String, BoundValue> option : options.entrySet()) {
+      BoundValue boundValue = option.getValue();
+      if (boundValue.isDefault()) {
+        continue;
+      }
+
+      for (Class<?> definingClass : keyToInterfaceMap.get(option.getKey())) {
+        Object value = boundValue.getValue() == null ? "" : boundValue.getValue();
+        builder.add(option.getKey(), value)
+           .withNamespace(definingClass);
+      }
+    }
+  }
+
+  private boolean canPopulateDisplayData() {
+    return jsonOptions.isEmpty();
+  }
+
+  private Multimap<String, Class<?>> buildOptionToInterfaceMap(
+      Set<PipelineOptionsReflector.Property> props) {
+    Multimap<String, Class<?>> keyToInterfaceMap = HashMultimap.create();
+    for (PipelineOptionsReflector.Property prop : props) {
+      keyToInterfaceMap.put(prop.name(), prop.definingInterface());
+    }
+
+    // Filter out overridden options
+    for (Map.Entry<String, Collection<Class<?>>>entry : keyToInterfaceMap.asMap().entrySet()) {
+      List<Class<?>> ifaces = Lists.newArrayList(entry.getValue());
+      for (int i = 0; i < ifaces.size(); i++) {
```

Comment:
Can you add some comments and/or extract some methods to explain these loops? 3 nested loops, including 2 that interact with their indices in interesting ways merits a description.